File: spoon-runner/src/test/java/com/squareup/spoon/SpoonCoverageMergerTest.java
package com.squareup.spoon;

import com.google.common.collect.ImmutableSet;
import java.io.File;
import java.io.IOException;
import java.util.Set;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.TemporaryFolder;

import static com.squareup.spoon.SpoonUtils.sanitizeSerial;
import static java.lang.String.format;
import static org.junit.Assert.assertTrue;

public final class SpoonCoverageMergerTest {

  @Rule
  public TemporaryFolder testFolder = new TemporaryFolder();

  @Test
  public void shouldMergeCoverageFiles() throws Exception {
    String serialId1 = "10.0.0.1:1234";
    String serialId2 = "10.0.0.2:1234";
    File spoonOutputDirectory = testFolder.newFolder("output");
    createTemporaryCoverageFiles(serialId1, serialId2);
    Set<String> serials = ImmutableSet.of(serialId1, serialId2);
    SpoonCoverageMerger spoonCoverageMerger = new SpoonCoverageMerger();

    spoonCoverageMerger.mergeCoverageFiles(serials, spoonOutputDirectory);

    File mergedCoverageFile = new File(spoonOutputDirectory, "/coverage/merged-coverage.ec");
    assertTrue(mergedCoverageFile.exists());
  }

  private void createTemporaryCoverageFiles(String serialId1, String serialId2) throws IOException {
    testFolder.newFolder("output", "coverage", sanitizeSerial(serialId1));
    testFolder.newFolder("output", "coverage", sanitizeSerial(serialId2));
    testFolder.newFile(format("output/coverage/%s/coverage.ec", sanitizeSerial(serialId1)));
    testFolder.newFile(format("output/coverage/%s/coverage.ec", sanitizeSerial(serialId2)));
  }
}

File: spoon-runner/src/test/java/com/squareup/spoon/SpoonCoverageMergerTest.kt
package com.squareup.spoon

import com.squareup.spoon.SpoonUtils.sanitizeSerial
import org.junit.Assert.assertTrue
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder
import java.io.File
import java.lang.String.format

class SpoonCoverageMergerTest {
  @JvmField @Rule val testFolder = TemporaryFolder()

  @Test fun shouldMergeCoverageFiles() {
    val outputDirectory = testFolder.newFolder("output")
    val serialId1 = "10.0.0.1:1234"
    val serialId2 = "10.0.0.2:1234"
    createTemporaryCoverageFiles(serialId1, serialId2)

    mergeCoverageFiles(setOf(serialId1, serialId2), outputDirectory)
    val mergedCoverageFile = File(outputDirectory, "/coverage/merged-coverage.ec")
    assertTrue(mergedCoverageFile.exists())
  }

  @Test fun shouldMergeMultiInstrumentedCoverageFiles() {
    val outputDirectory = testFolder.newFolder("coverage")
    val test1 = "class1_test1"
    val test2 = "class1_test2"
    val test3 = "class2_test1"
    createTemporaryMultiTestsCoverageFiles(test1, test2, test3)

    mergeAllCoverageFiles(outputDirectory)
    val mergedCoverageFile = File(outputDirectory, "coverage.ec")
    assertTrue(mergedCoverageFile.exists())
  }

  private fun createTemporaryCoverageFiles(serialId1: String, serialId2: String) {
    testFolder.newFolder("output", "coverage", sanitizeSerial(serialId1))
    testFolder.newFolder("output", "coverage", sanitizeSerial(serialId2))
    testFolder.newFile(format("output/coverage/%s/coverage.ec", sanitizeSerial(serialId1)))
    testFolder.newFile(format("output/coverage/%s/coverage.ec", sanitizeSerial(serialId2)))
  }

  private fun createTemporaryMultiTestsCoverageFiles(test1: String, test2: String, test3: String) {
    testFolder.newFile(format("coverage/%s-coverage.ec", test1))
    testFolder.newFile(format("coverage/%s-coverage.ec", test2))
    testFolder.newFile(format("coverage/%s-coverage.ec", test3))
  }
}
